File: coding-test/javascript/src/weekly-challenge/vowel-dictionary.js
// @ts-check

/**
 * @param {string} word
 * @returns {number}
 */
function solution(word) {
  let answer = 0
  let cnt = 3905
  for (let i = 0; i < word.length; i += 1) {
    answer += 'AEIOU'.indexOf(word.charAt(i)) * Math.floor((cnt /= 5)) + 1
  }

  return answer
}

/**
 * @param {string} word
 * @returns {number}
 */
function diffSolution(word) {
  return word
    .split('')
    .reduce(
      (cnt, c, i) =>
        cnt + 'AEIOU'.indexOf(c) * Math.floor(3905 / Math.pow(5, i + 1)) + 1,
      0
    )
}

const word = 'AAAAE'
console.log(solution(word))
console.log(diffSolution(word))

File: practice-coding-test/java-kotlin/src/kr/co/programmers/java/weekly_challenge/VowelDictionary.java
package kr.co.programmers.java.weekly_challenge;


import java.util.ArrayList;
import java.util.List;

public class VowelDictionary {

    public static void main(String[] args) {
        System.out.println(new Solution().solution("AAAAE"));
        System.out.println(new Solution().diffSolution("AAAAE"));
    }

    static class Solution {

        static List<String> list = new ArrayList<>();

        public int solution(String word) {
            dfs("", 0);
            return list.indexOf(word);
        }

        private void dfs(String str, int len) {
            if (len > 5) return;
            list.add(str);
            for (int i = 0; i < 5; i++) {
                dfs(str + "AEIOU".charAt(i), len + 1);
            }
        }

        public int diffSolution(String word) {
            int answer = 0;
            int per = 3905;
            for (String s : word.split("")) {
                answer += "AEIOU".indexOf(s) * (per /= 5) + 1;
            }
            return answer;
        }
    }
}
